Add example R script

jashapiro opened the pull request:
This PR adds an example R script that is designed mostly as an example of how a script might be structured to take advantage of command line options. 

The script takes a MAF file, tallies the counts of samples mutated and total mutations for each gene, and outputs a table with that information. There are a few options for filtering by variant frequency, etc. 

Originally there was to be an option to exclude commonly mutated genes, which I prepared for by including the `FLAGS.tsv` file in the reference directory. As I was writing though, I decided that might be more appropriate to use in the next planned step, a visualization notebook. There that information could be used for filtering, or for coloring plots.

The other script included is a script to generate MAF files separated by cancer type (short_histology, in this case) . The files are directly downloaded from the OpenPBT S3 bucket, and placed by default in `data/processed/maf_files`. I expect that we will want to include some of the resulting files in the repo, but probably not all, as some are large. In particular, the HGAT file  is ~87MB, as those samples have a lot of mutations (the others are <12 MB). I didn't include the results in the repo yet; I do think some of the files should be included, but maybe only smaller ones?

One decision I made when writing the script was to use `::` notation almost exclusively. The reasons were two: one for explicitness & clarity, and one to avoid annoying output every time the script runs. 

On the subject of annoying output, the MAF file reading could be quieted, but I wasn't sure whether to do that... On the flip side, I also thought about progress messages. For a quick script like this they seemed unnecessary, but including them can be nice (the setup script could benefit since reading and downloading the main MAF file can be slow). Curious to hear thoughts on that front.

Also curious to hear other enhancements that might come to mind!


closes #3

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env Rscript
```
May be worth pointing out (to emphasize further after scripting) that this path is not strictly universal, but also one doesn't have to rely on it.

jashapiro replied:
I expect to mention this in slides, yes. (This one is pretty universal though!)

sjspielman replied:
Mine's in `/usr/local/bin/`, is the `env` from a conda installation or similar? Just curious!

jashapiro replied:
That is odd! Do you maybe have it in both places and `local` is first in your PATH? As far as conda, the version installed with conda will be somewhere else, but the way `env` works it should still launch commands from the _current_ environment, regardless of which environment the `env` program resides in. For example, when I run `/usr/bin/env python` I get whatever version of python is active in my conda environment, not the base python (same as if I ran `env python`

sjspielman replied:
If you thought that was odd, wait til you read that... I do not have a `/usr/bin/env/` directory at all 
<img width="467" alt="Screen Shot 2022-05-23 at 4 48 38 PM (2)" src="https://user-images.githubusercontent.com/4701111/169903671-104c051b-f71b-4e4b-95ab-10808beca163.png">

jashapiro replied:
`/usr/bin` is the directory, which I really hope you have! `env` should be inside it, but is a binary (or maybe link)

sjspielman replied:
Lord help me I absolutely have been reading this _without a space_ as `usr/bin/env/Rscript`. 
I have it all, it just loses to `/usr/local/bin` in my `$PATH`

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+# Option descriptions:
+#
+# --maf :  File path to MAF file to be analyzed. Can be .gz compressed.
+# --outfile : The name of the output file to create
```
```suggestion
# --outfile : The path of the output file to create
```
This may not be clear to them otherwise that name=path

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+
+# Example invocation:
+#
+# Rscript count_gene_mutations.R \
```
Maybe also show a 1-line version of this? I am not suggesting to delete this, but to complement with a version `\`-free since separating onto new lines may be new for them. 
That said, we probably want them to get in the habit of writing clearly across multiple lines, so I don't want to add a one-liner and lose that message.

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+# Parse options
+opts <- parse_args(OptionParser(option_list = option_list))
+
+# Check input files
```
```suggestion
# Check that the input file paths are correctly specified
```

More verbose for trainees

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+# Get the excluded genes list if provided
+if(!is.null(opts$exclude_genes)){
+  exclude_genes <- readr::read_tsv(opts$exclude_file) %>%
+    pull("gene")
```
```suggestion
    dplyr::pull("gene")
```

sjspielman replied:
Quotes also not needed here because of tidyselect magic, but doesn't matter.

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+
+# Define the MAF `Consequence` values we are interested in and their classification
+#  based on definitions in http://asia.ensembl.org/Help/Glossary?id=535
+syn_class <- c(
```
This is a good teachable moment - spreading code out onto multiple lines makes it easier to read, write, reuse, etc.

sjspielman commented on scripts/count_gene_mutations.R:
```diff
@@ -0,0 +1,154 @@
+
+  
+# Write output
+readr::write_tsv(gene_counts, file = opts$outfile)
```
EOF line

sjspielman commented on data/reference/README.md:
```diff
@@ -0,0 +1,5 @@
+# Reference directory files and origins
+
+- `FLAGS.tsv`:
+The FLAGS.tsv file is a list of commonly mutated genes in exome sequencing samples, derived from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267152/.
```
```suggestion
The `FLAGS.tsv` file is a list of commonly mutated genes in exome sequencing samples, derived from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267152/.
```

sjspielman commented on data/reference/README.md:
```diff
@@ -0,0 +1,5 @@
+
+- `FLAGS.tsv`:
+The FLAGS.tsv file is a list of commonly mutated genes in exome sequencing samples, derived from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267152/.
+This file contains the top 50 genes from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5706417/bin/12920_2017_309_MOESM3_ESM.txt
```
I think it might be worthwhile to create this file with a shell script - `curl` and `head -n 50 > FLAGS.tsv`. It would jump off very nicely from the fastq activity as a concrete example.

jashapiro replied:
I was debating this, so I'll go ahead and do it. I honestly didn't remember how much processing was involved, but not much.

sjspielman replied:
Script looks good!

sjspielman reviewed (commented):
I think this is a good scaffold to show them. One of the suggestions is to make a comment a little more verbose, and I think overall more commenting throughout would be beneficial. For example, more direct indications for what `stop()` is doing. This is especially important since we're not exactly going go line by line through what this code does.

I also think we need a rename to `00-count_gene_mutations.R`?

jashapiro reviewed (commented):
> For example, more direct indications for what `stop()` is doing. This is especially important since we're not exactly going go line by line through what this code does.

This was a debate I definitely had... How many teaching comments to make. I ultimately decided to go a bit sparser, because the goal here was not necessarily to teach R, but to give somewhat realistic example.

In the case of `stop()` I was planning to explain that in slides, in the context of "check your files and parameters first, before you do anything that might take a long time". I could add a brief note to that effect in the comments

For reference, I was also planning to demonstrate failures, and what you get "for free" `optparse`, but incorporating that content into comments is hard...

sjspielman commented:
> I ultimately decided to go a bit sparser, because the goal here was not necessarily to teach R, but to give somewhat realistic example.

I don't expect to be teaching R, but you've written a pretty solid script with good lessons in it for them to take away with them for their future R coding! I view comments here as a resource for them to look more over this script later if they want, but definitely not something to be explaining in-depth (or frankly at all?) during the workshop. 

The other thought I had is an R-specific thought. I personally find the [RStudio comment feature](https://style.tidyverse.org/files.html?q=comment#internal-structure) where it sees `# comment ----` as a "header" and pops it into the "table of contents" to be _extremely_ useful. It really helps me personally navigate my code, especially when the code gets long. This structure could be worth incorporating, but again we don't want to be teaching R (or necessarily RStudio tricks).

jashapiro commented:
I updated the comments some, including adding the RStudio header marks, which I was vaguely aware of, but didn't know about specifically! If there are still more comments that you think would be helpful, feel free to suggest them. 

I also added here an Rproject file, with settings I like. We can talk about some/all of those settings when introducing this w/ slides. (One note, there is now a setting to use One Sentence Per Line with the Rmarkdown graphical editor, which makes it something that I could see actually using, but still won't mostly... unless references?)

I also added a set of data files, excluding the one (HGAT) that was 90 MB.

sjspielman commented on data/reference/README.md:
```diff
@@ -0,0 +1,6 @@
+- `FLAGS.tsv`:
+The `FLAGS.tsv` file is a list of commonly mutated genes in exome sequencing samples, derived from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4267152/.
+This file contains the top 50 genes from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5706417/bin/12920_2017_309_MOESM3_ESM.txt
+See `scripts/setup/download_flags.sh` for details about this file was created.
```
```suggestion
See `scripts/setup/download_flags.sh` for details about how this file was created.
```

sjspielman commented on rrp-workshop-exercises.Rproj:
```diff
@@ -0,0 +1,17 @@
+Version: 1.0
+
+RestoreWorkspace: No
+SaveWorkspace: No
```
🎉

sjspielman reviewed (commented):
In the interest of emphasizing the importance to documentation, I would suggest adding a `README` in `scripts/setup/`. Doesn't need to be anything beyond a bullet list with a quick description of what each of the two scripts does and how to run them.

jashapiro commented:
Okay, I think this is ready for one more look! I added a few more "teaching" comments about the option parsing and what `stop()` does.

I also added a README to the setup folder, which I definitely think should be there. Thank you for catching that!
